lib/mocha-headless-chrome.js

```diff
@@ -0,0 +1,229 @@
+      M.reporters.Base.call(this, runner)
+    }
+  }
+
+  Object.defineProperty(window, 'mocha', {
+    get: function() {
+      return undefined
+    },
+    set: function(m) {
+      shimMochaInstance(m)
+      delete window.mocha
+      window.mocha = m
+    },
+    configurable: true,
+  })
+
+  Object.defineProperty(window, 'Mocha', {
+    get: function() {
+      return undefined
+    },
+    set: function(m) {
+      shimMochaProcess(m)
+      delete window.Mocha
+      window.Mocha = m
+    },
+    configurable: true,
+  })
+}
+
+function configureViewport(width, height, page) {
+  if (!width && !height) return page
+
+  let viewport = page.viewport()
+  width && (viewport.width = width)
+  height && (viewport.height = height)
+
+  return page.setViewport(viewport).then(() => page)
+}
+
+function handleConsole(msg) {
+  const args = msg._args
+
+  Promise.all(args.map(a => a.jsonValue())).then(args => {
+    // process stdout stub
+    let isStdout = args[0] === 'stdout:'
+    isStdout && (args = args.slice(1))
+    //
+    let msg = util.format(...args)
+    !isStdout && (msg += '\n')
+    process.stdout.write(msg)
+  })
+}
+
+function prepareUrl(filePath) {
+  if (/^[a-zA-Z]+:\/\//.test(filePath)) {
+    // path is URL
+    return filePath
+  }
+
+  // local path
+  let resolvedPath = path.resolve(filePath)
+  return `file://${resolvedPath}`
+}
+
+exports.runner = function({
+  file,
+  reporter,
+  timeout,
+  width,
+  height,
+  args,
+  executablePath,
+  visible,
+}) {
+  return new Promise(resolve => {
+    // validate options
+    if (!file) {
+      throw new Error('Test page path is required.')
+    }
+
+    args = [].concat(args || []).map(arg => '--' + arg)
+    !timeout && (timeout = 60000)
+
+    const url = prepareUrl(file)
+
+    const options = {
+      ignoreHTTPSErrors: true,
+      headless: !visible,
+      executablePath,
+      args,
+    }
+
+    const result = puppeteer.launch(options).then(browser =>
+      browser
+        .pages()
+        .then(pages => pages.pop())
+        .then(configureViewport.bind(this, width, height))
+        .then(page => {
+          page.on('console', handleConsole)
+          page.on('dialog', dialog => dialog.dismiss())
+          //page.on('pageerror', err => console.error(err));
+          page.on('pageerror', err => {
+            page.evaluate(message => {
+              window.onerror(new Error(message))
+            }, err.message)
+          })
+
+          return page
+            .evaluateOnNewDocument(initMocha, reporter)
+            .then(() => page.goto(url))
+            .then(() =>
+              page.waitForFunction(() => window.__mochaResult__, {
+                timeout: timeout,
+              })
+            )
+            .then(() => page.evaluate(() => window.__mochaResult__))
+            .then(obj => {
+              browser.close()
+              return obj
+            })
+        })
+    )
+
+    resolve(result)
```

Comment:
❓ Would resolving the promise ensure that, beyond this point, the data is accessible? Is this a synchronization point?